ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -0,0 +1,128 @@
+module Codec.Binary.Bech32
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  , Word5(UnsafeWord5)
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor, testBit)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8, Word32)
+import Data.Hashable (Hashable(..))
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+newtype Word5 = UnsafeWord5 Word8
+              deriving (Eq)
+
+instance Hashable Word5 where
+    hashWithSalt salt (UnsafeWord5 x) = hashWithSalt salt x
+
+word5 :: Word8 -> Word5
+word5 x = UnsafeWord5 (x .&. 31)
+
+fromWord5 :: Num a => Word5 -> a
+fromWord5 (UnsafeWord5 x) = fromIntegral x
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
```

Comment:
I have pushed a commit that does this just to see. Could you explain why this is better? Your `charsetMap` is the equivalent of a the Data.HashMap. Access is constant time in either case (the Hashable instance of Char is just `fromEnum`, so no collisions there either). Memory usage is similar. But now, there is this extra non-trivial code. The HashMap version was very easy to read.

Would you be much opposed to just leaving the HashMap version as it was? This might be premature optimization. There is much lower hanging fruit in the base32->base256 conversion.